### ISSUE TITLE
assertions in RefineMesh do not pass

### ISSUE THREAD

**Issue by raphaelsulzer** [cdcseacave/openMVS#552]
Hello, and thank you for the great software!

### Specifications like the version of the project, operating system, and hardware
-Ubuntu 18.04
-CUDA support disabled in cmake because it was causing issues
-Running the current master branch of openMVS 

### Steps to reproduce the problem
I ran a pipeline with a DTU benchmark datasete which runs fine up until the RefineMesh step.
Here I get the following output:

Debugging starts
18:49:20 [App     ] Build date: Apr 23 2020, 18:31:21
18:49:20 [App     ] CPU: Intel(R) Core(TM) i5-6440HQ CPU @ 2.60GHz (4 cores)
18:49:20 [App     ] RAM: 7.66GB Physical Memory 10.25GB Virtual Memory
18:49:20 [App     ] OS: Linux 5.3.0-46-generic (x86_64)
18:49:20 [App     ] SSE & AVX compatible CPU & OS detected
18:49:20 [App     ] Command line: -w /home/adminlocal/PhD/data/DTU_Jensen/scene6/openMV/ -i scene6_mesh.mvs -o scene6_mesh_refined.ply
18:49:36 [App     ] Scene loaded (15s570ms):
	49 images (49 calibrated) with a total of 89.72 MPixels (1.83 MPixels/image)
	0 points, 1047599 vertices, 2094718 faces
18:49:36 [App     ] Refine mesh at: 0.25 image scale
Decimated faces 9170 (0.49%, 8s, ETA 0ms)...
...
Decimated faces 1885246 (100%, 39s398ms)          

18:52:22 [App     ] Mesh decimated: 2094718 -> 209472 faces
18:52:23 [App     ] Removed 0 zero-area faces
18:52:24 [App     ] Removed 0 duplicate faces
18:52:24 [App     ] Removed 0 non-manifold faces
18:52:24 [App     ] Removed 0 degenerate vertices
18:52:24 [App     ] Removed 45 unreferenced vertices
18:52:25 [App     ] Removed 0 duplicate vertices
18:52:25 [App     ] Split 0 non-manifold vertices
18:52:26 [App     ] Closed 0 holes and added 0 new faces
18:52:26 [App     ] Cleaned mesh: 104976 vertices, 209472 faces (45s159ms)
18:52:27 [App     ] Removed 0 zero-area faces
18:52:27 [App     ] Removed 0 duplicate faces
18:52:27 [App     ] Removed 0 non-manifold faces
18:52:27 [App     ] Removed 0 degenerate vertices
18:52:27 [App     ] Removed 0 unreferenced vertices
18:52:27 [App     ] Removed 0 duplicate vertices
18:52:27 [App     ] Split 0 non-manifold vertices
18:52:28 [App     ] Closed 0 holes and added 0 new faces
18:52:28 [App     ] Removed 0 non-manifold faces
18:52:28 [App     ] Removed 0 non-manifold vertices
18:52:28 [App     ] Cleaned mesh: 104976 vertices, 209472 faces (2s83ms)

RefineMesh: /home/adminlocal/PhD/cpp/openMVS/libs/MVS/Mesh.cpp:2998: void MVS::Mesh::Subdivide(const AreaArr&, uint32_t): Assertion `faces.GetSize()-(faces.GetCapacity()/3) > mapSplits.size()' failed.

if I comment this assertion, I get a failure here:
RefineMesh: /home/adminlocal/PhD/cpp/openMVS/libs/MVS/SceneRefine.cpp:121: void MeshRefine::RasterMesh::Raster(const ImageRef&): Assertion `z > 0' failed.

if I also comment this, I seem to get a bit further but another assertion fails later:
Processed iterations 0 (0.00%, 1m36s, ETA 0ms)...
Processed iterations 1 (4.55%, 3m12s, ETA 1h7m)...
Processed iterations 2 (9.09%, 4m48s, ETA 48m)... 
Processed iterations 3 (13.64%, 6m24s, ETA 40m)...
Processed iterations 4 (18.18%, 8m3s, ETA 36m)... 
Processed iterations 5 (22.73%, 9m39s, ETA 32m)...
Processed iterations 6 (27.27%, 11m17s, ETA 30m)...
Processed iterations 7 (31.82%, 12m59s, ETA 27m)...
Processed iterations 8 (36.36%, 14m41s, ETA 25m)...
Processed iterations 9 (40.91%, 16m21s, ETA 23m)...
Processed iterations 10 (45.45%, 18m2s, ETA 21m)...
Processed iterations 11 (50.00%, 19m47s, ETA 19m)...
Processed iterations 12 (54.55%, 21m24s, ETA 17m)...
Processed iterations 13 (59.09%, 23m1s, ETA 15m)... 
Processed iterations 14 (63.64%, 24m38s, ETA 14m)...
Processed iterations 15 (68.18%, 26m16s, ETA 12m)...
Processed iterations 16 (72.73%, 27m52s, ETA 10m)...
Processed iterations 17 (77.27%, 29m28s, ETA 8m)... 
Processed iterations 18 (81.82%, 31m5s, ETA 6m)... 
Processed iterations 19 (86.36%, 32m44s, ETA 5m)...
Processed iterations 20 (90.91%, 34m29s, ETA 3m)...
Processed iterations 22 (100%, 36m7s502ms)        

20:08:08 [App     ] Mesh saved: 88270 vertices, 175700 faces (46ms)
20:08:08 [App     ] Refine mesh at: 1.00 image scale

RefineMesh: /home/adminlocal/PhD/cpp/openMVS/libs/MVS/SceneRefine.cpp:480: void MeshRefine::ListFaceAreas(MVS::Mesh::AreaArr&): Assertion `(idxFace == std::numeric_limits<uint32_t>::max() && views[idxImage].depthMap(j,i) == 0) || (idxFace != std::numeric_limits<uint32_t>::max() && views[idxImage].depthMap(j,i) > 0)' failed.

Does anyone have an idea what is the problem here?

**Comment by cdcseacave** [comment]
I can have a look into them, but these ASSERTS might not be well tested, so not sure how well they need to be enforced.

However, you should definitely use the software in Release mode, in order to get huge gaines in speed. This way all the asserts will be omitted as well, and you can see if the result looks good or if there is really a problem with the code.

**Comment by raphaelsulzer** [comment]
Yes, I was in debug mode to try to find a solution for the cuda error which is mentioned in some other issues (the software fails with cuda error, even when "--use-cuda 0" is passed). Then I noticed these failures. My result after commenting the asserts, or running in release mode look fine. I'll leave it up to you to close this.
Thanks again!

**Comment by cdcseacave** [comment]
Sorry to hear you have CUDA issue. Unfortunately I do not have a nvidia card anymore so I can not help with this, and so I didn't test this mode since long time.

I'll let this open, as maybe this asserts will prove to be important, so we will need to fix it. Or somebody will help change the asserts to fit.

**Comment by cdcseacave** [comment]
Closing as stale. The codebase has been substantially updated on the \develop\ branch. Please reopen with reproduction steps if the issue persists on the latest code.
